test/borda_clj/trace.clj

```diff
@@ -0,0 +1,196 @@
+             [clojure.spec.alpha :as s]
+             [clojure.spec.gen.alpha :as gen]
+             [clojure.spec.test.alpha :as stest]
+             [borda-clj.trace :as trace]
+             [clojure.pprint :as pp :refer (pprint)]))
+
+(defn run-trace [global-dimensions max-buffer-size trace]
+  (let [[report-ch measurements-ch kill-ch timeout-ch trace-ch] (repeatedly chan)
+        successful-send (fn [m] (>!! trace-ch {::send m}))
+        failed-send (fn [& args] (throw (RuntimeException. "boom")))
+        send (fn [measurements] ((or (:trace-metadata (meta measurements))
+                                    ;; it doesn't matter much whether we succeed
+                                    ;; or fail in the last sending attempt,
+                                    ;; since we never retry that one. it
+                                    ;; slightly simplifies tests to assume it
+                                    ;; succeeds, so let's do that.
+                                    successful-send)
+                                measurements))
+        on-send-error (fn [m e]
+                        ;;(clojure.pprint/pprint e)
+                        (>!! trace-ch {::error m}))
+        flush-ch (sut/go-flush-measurements report-ch measurements-ch kill-ch global-dimensions ::bogus-interval send on-send-error (constantly timeout-ch))]
+    (sut/go-reduce-reports report-ch measurements-ch kill-ch max-buffer-size)
+    (go
+      (<! flush-ch)
+      (close! trace-ch))
+    (go
+      (doseq [op trace]
+        (case op
+          :successful-tick (>! timeout-ch successful-send)
+          :failed-tick (>! timeout-ch failed-send)
+          :delay (<! (timeout 100))
+          (>! report-ch op)))
+      (<! (timeout 100))
+      (close! report-ch))
+    (<!! (async/into [] trace-ch))))
+
+(s/def ::key simple-keyword?)
+(s/def ::finite-number (s/or :int int? :double (s/and double? #(Double/isFinite %))))
+(s/def ::dimensions (s/every-kv ::key
+                                (s/or :str (s/and string? not-empty) :num ::finite-number)
+                                :min-count 1))
+(s/def ::values (s/every-kv ::key ::finite-number
+                            :min-count 1))
+(s/def ::measurement (s/tuple ::dimensions ::values))
+(s/def ::measurements (s/map-of ::dimensions ::values))
+(s/def ::trace-input-entry (s/or :kw #{:successful-tick :failed-tick :delay} :measurement ::measurement))
+(s/def ::input-trace (s/coll-of ::trace-input-entry :max-count 10))
+(s/def ::send ::measurements)
+(s/def ::final-send ::measurements)
+(s/def ::error ::measurements)
+(s/def ::trace-output-entry (s/and (s/or :send (s/keys :req [::send])
+                                         :error (s/keys :req [::error]))
+                                   ;; this predicate runs on the conformed value
+                                   #(= (count (second %)) 1)))
+(s/def ::output-trace (s/coll-of ::trace-output-entry))
+
+(defn gen-run-trace-args
```

Comment:
This is tricky stuff and it can probably be made much simpler and improved in other ways; I was just experimenting.